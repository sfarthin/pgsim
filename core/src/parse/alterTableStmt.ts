import {
  array,
  object,
  Decoder,
  number,
  constant,
  either4,
  pojo,
} from "decoders";
import {
  Relation,
  relationDecoder,
  ColumnDef,
  columnDefDecoder,
} from "./createStmt";
import { Constraint, constraintDecoder } from "./constraint";

export enum AlterTableCmdSubType {
  ADD = 0,
  ALTER_COLUMN = 3,
  DROP = 10,
  ADD_CONSTRAINT = 14,
}

export type AlterTableCmd =
  | {
      subtype: AlterTableCmdSubType.ADD;
      def: {
        ColumnDef: ColumnDef;
      };
      behavior: number;
    }
  | {
      subtype: AlterTableCmdSubType.DROP;
      behavior: number;
    }
  | {
      subtype: AlterTableCmdSubType.ADD_CONSTRAINT;
      def: {
        Constraint: Constraint;
      };
      behavior: number;
    }
  | {
      subtype: AlterTableCmdSubType.ALTER_COLUMN;
      name: string;
      def: any;
      behavior: number;
    };

export const alterTableCmdDecoder: Decoder<AlterTableCmd> = either4(
  object({
    subtype: constant(0),
    def: object({
      ColumnDef: columnDefDecoder,
    }),
    behavior: number,
  }),
  object({
    subtype: constant(3),
    def: pojo,
    behavior: number,
  }),
  object({
    subtype: constant(10),
    behavior: number,
  }),
  object({
    subtype: constant(14),
    def: object({
      Constraint: constraintDecoder,
    }),
    behavior: number,
  })
);

export type AlterTableStmt = {
  relation: Relation;
  cmds: Array<{ AlterTableCmd: AlterTableCmd }>;
  relkind: number;
};

export const alterTableStmtDecoder: Decoder<AlterTableStmt> = object({
  relation: relationDecoder,
  cmds: array(object({ AlterTableCmd: alterTableCmdDecoder })),
  relkind: number,
});
