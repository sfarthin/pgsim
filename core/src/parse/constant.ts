import { Decoder, object, string, number, either3, either } from "decoders";

export type StringValue = { str: string };
export const stringDecoder: Decoder<StringValue> = object({ str: string });
export type FloatValue = { str: string };
export const floatDecoder: Decoder<FloatValue> = object({ str: string });
export type IntegerValue = { ival: number };
export const integerDecoder: Decoder<IntegerValue> = object({ ival: number });

export type Value = StringValue | FloatValue | IntegerValue;
export const valueDecoder: Decoder<Value> = either3(
  stringDecoder,
  floatDecoder,
  integerDecoder
);

export type Constant = {
  val: StringValue | FloatValue | IntegerValue;
};
export const constantDecoder: Decoder<Constant> = object({
  val: either3(stringDecoder, floatDecoder, integerDecoder),
});

export type A_Const = {
  val:
    | { Float: FloatValue }
    | { String: StringValue }
    | { Integer: IntegerValue };
};

export const aConstDecoder = object({
  val: either3(
    object({ Float: floatDecoder }),
    object({ String: stringDecoder }),
    object({ Integer: integerDecoder })
  ),
});
