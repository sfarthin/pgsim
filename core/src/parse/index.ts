import { optional, mixed, array, string, guard, exact } from "decoders";
import pgParse from "./pgParse";
import { queryDecoder, Queries } from "./query";
import toOriginalText from "../toReader/toOriginalText";
import { SelectStmt } from "./selectStmt";
import { PGError, PGErrorCode } from "../errors";

export * from "./selectStmt";
export * from "./alterTableStmt";
export * from "./createStmt";
export * from "./constraint";
export * from "./query";
export * from "./constant";

export const parserResultDecoder = exact({
  // This is mixed because Error messages are hard to read if we do this here, we validate each query seperately
  query: array(mixed),
  stderr: optional(string),
  error: optional(mixed),
});

export const validateParsedResult = guard(parserResultDecoder);

export const validateQuery = guard(queryDecoder);

const getLocationFromUnSafeQuery = (q: any) => {
  // @ts-ignore - with the conditional chaining this is harmless, byt typescript is concerned
  return Object.values(q)?.[0]?.relation?.RangeVar?.location ?? null;
};

export default function parse(sql: string): Queries {
  const unsafeResult = pgParse(sql);
  const { query: queries, stderr, error } = validateParsedResult(unsafeResult);

  // Native error
  if (stderr || error) {
    throw error || new Error(stderr);
  }

  let index = 0;
  try {
    return queries.map((unSafeQuery: any, i) => {
      index = i;
      return validateQuery(unSafeQuery);
    });
  } catch (e) {
    try {
      const queryAsText = toOriginalText(sql, index);
      e.message = `Unable to understand query ${`\n\n${queryAsText}\n\n`} ${
        e.message
      }`;
    } catch (e) {
      e.message = `Unable to understand query ${index + 1} ${e.message}`;
    }
    throw e;
  }
}

export function parseSelect(str: string): SelectStmt {
  const queries = parse(str);

  if (queries.length !== 1 || !("SelectStmt" in queries[0])) {
    throw new PGError(PGErrorCode.NOT_EXISTS, "Expected a single SELECT");
  }

  return queries[0].SelectStmt;
}
