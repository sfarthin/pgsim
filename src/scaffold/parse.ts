import { resolve } from "path";
import { writeFileSync, readdirSync, existsSync } from "fs";
import prettier from "prettier";
import warning from "./warning";
import { stmtTypes, mapStmtTypes } from "./constants";
import { camelCase } from "lodash";

function mapStmtChunks(fn: (i: string[]) => string) {
  const chunkSize = 10;

  const chunks = [];

  for (let i = 0; i < stmtTypes.length; i += chunkSize) {
    chunks.push(stmtTypes.slice(i, i + chunkSize));
    // do whatever
  }

  return chunks.map((chunk) => fn(chunk)).join("\n");
}

export default async function () {
  const parseStmtPath = resolve(__dirname, "../parse/stmt.ts");
  const parseIndexPath = resolve(__dirname, "../parse/index.ts");

  const files = readdirSync(resolve(__dirname, "../parse/")).flatMap((f) =>
    f === "index.ts" ? [] : f.substring(0, f.length - 3)
  );

  const mapFiles = (mapFn: (n: string) => string) =>
    files.map(mapFn).join(`\n`);

  const importedParsers = mapStmtTypes(
    (stmtName) =>
      `import { ${camelCase(stmtName)} } from "./${camelCase(stmtName)}";`
  );

  const exportedStmt = mapFiles((stmtName) => `export * from "./${stmtName}";`);

  writeFileSync(
    parseStmtPath,
    prettier.format(
      /* ts */ `

    ${warning}
    import { or, addStmtType } from "./util";
    ${importedParsers}

    export const stmt = or([
      ${mapStmtChunks(
        (chunk) => /* ts */ `or([
          ${chunk.map((f) => `addStmtType('${f}', ${camelCase(f)})`).join(",")}
        ]),`
      )}
    ]);

    `,
      { parser: "babel" }
    )
  );

  writeFileSync(
    parseIndexPath,
    prettier.format(
      /* ts */ `

    ${warning}
    import { parse } from "./stmts";
    export default parse;
    ${exportedStmt}

    `,
      { parser: "babel" }
    )
  );

  for (const stmtType of stmtTypes) {
    const stmtFilePath = resolve(
      __dirname,
      `../parse/${camelCase(stmtType)}.ts`
    );
    if (!existsSync(stmtFilePath)) {
      writeFileSync(
        stmtFilePath,
        prettier.format(
          /* ts */ `
        import { ${stmtType} } from "~/types";
        import { Rule, EOS, fail } from "./util";

        export const ${camelCase(stmtType)}:Rule<{
          value: { ${stmtType}: ${stmtType} };
          eos: EOS;
        }> = fail;
      `,
          { parser: "babel" }
        )
      );
    }
  }
}
