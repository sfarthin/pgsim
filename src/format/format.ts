import { Stmt } from "~/types";
import parse from "../parse";
import { toSingleLineIfPossible } from "./util";
import { toString, PrintOptions, createFriendlyStmtError } from "./print";
import { SuccessResult } from "../parse/util";
import { assertError } from "../assertError";
import formatStmt from "./stmt";

/**
 * If we already have the parse result, we can
 * pass that in rather than re-parsing it.
 */
export default function format(
  input: SuccessResult<Stmt[]>,
  filename: string,
  opts: Omit<PrintOptions, "filename">
): string;
export default function format(
  input: string,
  filename: string,
  opts?: Omit<PrintOptions, "sql" | "filename">
): string;
export default function format(
  input: string | SuccessResult<Stmt[]>,
  filename = "unknown",
  opts?: Partial<PrintOptions>
): string {
  const result =
    typeof input === "string" ? parse(input, filename, opts) : input;
  const sql = typeof input === "string" ? input : opts?.sql ?? "";

  const codeBlock = result.value.flatMap((stmt, i) => {
    try {
      const formattedStmt = toSingleLineIfPossible(formatStmt(stmt));
      if (i === 0) {
        return formattedStmt;
      } else {
        // Lets add a newline between statements.
        return [[], ...formattedStmt];
      }
    } catch (e) {
      assertError(e);

      throw createFriendlyStmtError(stmt, e, {
        sql,
        filename,
        colors: true,
        ...opts,
      });
    }
  });

  return toString(codeBlock, {
    sql,
    filename,
    ...opts,
  });
}
