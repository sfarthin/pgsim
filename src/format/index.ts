import createStmt from "./createStmt";
import variableSetStmt from "./variableSetStmt";
import createEnumStmt from "./createEnumStmt";
import createSeqStmt from "./createSeqStmt";
import alterSeqStmt from "./alterSeqStmt";
import dropStmt from "./dropStmt";
import alterEnumStmt from "./alterEnumStmt";
import alterOwnerStmt from "./alterOwnerStmt";
import alterTableStmt from "./alterTableStmt";
import viewStmt from "./viewStmt";
import selectStmt from "./selectStmt";
import renameStmt from "./renameStmt";
import comment from "./comment";
import updateStmt from "./updateStmt";
import { Stmt, StatementType } from "../types";
import {
  toLineAndColumn,
  getSnippetWithLineNumbers,
  findNextToken,
} from "../parse/error";
import indexStmt from "./indexStmt";
import parse from "../parse";
import c from "ansi-colors";
import { NEWLINE, textFormatter, Formatter } from "./util";

type Opts = {
  ignore?: StatementType[];
  ignoreAllExcept?: StatementType[];
  sql?: string; // <-- Original sql before formatting. useful when there is an error.
  filename?: string;
};

// const schemaStatements: StatementType[] = [
//   "CreateStmt",
//   "AlterTableStmt",
//   "CreateSeqStmt",
//   "VariableSetStmt",
//   "CreateEnumStmt",
//   "AlterSeqStmt",
//   "IndexStmt",
//   "CommentStmt",
// ];

function formatStmt<T>(stmt: Stmt, f: Formatter<T>, opts?: Opts): T[][] {
  const s = stmt.stmt;
  const statementType = Object.keys(s)[0] as StatementType;

  if (opts?.ignore && opts.ignore.includes(statementType)) {
    return comment(`-- Ignoring ${statementType} statements`, f);
  }

  if (opts?.ignoreAllExcept && !opts.ignoreAllExcept.includes(statementType)) {
    return comment(`-- Ignoring ${statementType} statements`, f);
  }

  if ("CreateStmt" in s) {
    return createStmt(s.CreateStmt, f);
  } else if ("DropStmt" in s) {
    return dropStmt(s.DropStmt, f);
  } else if ("VariableSetStmt" in s) {
    return variableSetStmt(s.VariableSetStmt, f);
  } else if ("CreateEnumStmt" in s) {
    return createEnumStmt(s.CreateEnumStmt, f);
  } else if ("CreateSeqStmt" in s) {
    return createSeqStmt(s.CreateSeqStmt, f);
  } else if ("AlterSeqStmt" in s) {
    return alterSeqStmt(s.AlterSeqStmt, f);
  } else if ("AlterEnumStmt" in s) {
    return alterEnumStmt(s.AlterEnumStmt, f);
  } else if ("AlterOwnerStmt" in s) {
    return alterOwnerStmt(s.AlterOwnerStmt, f);
  } else if ("IndexStmt" in s) {
    return indexStmt(s.IndexStmt, f);
  } else if ("AlterTableStmt" in s) {
    return alterTableStmt(s.AlterTableStmt, f);
  } else if ("Comment" in s) {
    return comment(s.Comment, f);
  } else if ("SelectStmt" in s) {
    return selectStmt(s.SelectStmt, f);
  } else if ("ViewStmt" in s) {
    return viewStmt(s.ViewStmt, f);
  } else if ("RenameStmt" in s) {
    return renameStmt(s.RenameStmt, f);
  } else if ("UpdateStmt" in s) {
    return updateStmt(s.UpdateStmt, f);
  }

  throw new Error(`Cannot format ${Object.keys(s)}`);
}

export default function format(_stmts: Stmt[] | string, opts?: Opts): string {
  const stmts = typeof _stmts === "string" ? parse(_stmts) : _stmts;
  const f = textFormatter;

  if (stmts.length === 0) {
    return "";
  }

  return stmts
    .map((stmt) => {
      try {
        return f.concat(formatStmt(stmt, f, opts));
      } catch (_e) {
        const e = _e as { name?: string; message?: string };
        const errorType = e.name === "Error" ? "" : `(${e.name})`;
        if (opts?.filename && opts?.sql) {
          // Lets skip over any comments
          const { start: pos } = findNextToken(
            opts.sql,
            stmt.stmt_location ?? 0
          );

          const { line, column } = toLineAndColumn(opts.sql, pos);

          e.name = `Problem formatting${errorType} ${c.cyan(
            opts.filename
          )}(${c.cyan(String(line + 1))},${c.cyan(String(column + 1))})`;

          e.message = `${
            e.message
          }${NEWLINE}${NEWLINE}${getSnippetWithLineNumbers({
            str: opts.sql,
            start: pos,
            end: (stmt.stmt_location ?? 0) + (stmt.stmt_len ?? 20),
          })}${NEWLINE}`;
        } else {
          e.name = `Problem formatting${errorType}`;
          e.message = `${e.message}${NEWLINE}${NEWLINE}${JSON.stringify(
            stmt,
            null,
            2
          )}`;
        }
        throw e;
      }
    })
    .join(`${NEWLINE}${NEWLINE}`);
}
