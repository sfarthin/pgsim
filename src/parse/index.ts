import { Stmt } from "../types";
import {
  transform,
  sequence,
  zeroToMany,
  whitespace,
  sqlStyleComment,
  cStyleComment,
  or,
  ResultType,
  Rule,
  Context,
  combineComments,
  endOfInput,
  FailResult,
  SuccessResult,
} from "./util";
import { getFriendlyErrorMessage } from "./error";
import { variableSetStmt } from "./variableSetStmt";
import { createEnumStmt } from "./createEnumStmt";
import { createSeqStmt } from "./createSeqStmt";
import { alterSeqStmt } from "./alterSeqStmt";
import { createStmt } from "./createStmt";
import { dropStmt } from "./dropStmt";
import { alterEnumStmt } from "./alterEnumStmt";
import { alterOwnerStmt } from "./alterOwnerStmt";
import { indexStmt } from "./indexStmt";
import { alterTableStmt } from "./alterTableStmt";
import { selectStmt } from "./selectStmt";
import { viewStmt } from "./viewStmt";
import { codeComments } from "./codeComments";
import { renameStmt } from "./renameStmt";
import { updateStmt } from "./updateStmt";

class ParseError extends Error {
  expectedStmt?: Stmt;
  // context: Context;
  // result: FailResult;
  constructor(context: Context, result: FailResult, expectedStmt?: Stmt) {
    super();
    this.message = getFriendlyErrorMessage(context, result);
    // this.context = context;
    // this.result = result;
    this.expectedStmt = expectedStmt;
  }

  toString() {
    return "error";
  }
}

const CommentStatement = transform(
  sequence([
    zeroToMany(whitespace),
    or([cStyleComment, sqlStyleComment]),
    zeroToMany(whitespace),
  ]),
  (v) => {
    return { value: { Comment: v[1] }, eos: null };
  }
);

export const stmt = transform(
  or([
    variableSetStmt,
    createEnumStmt,
    createSeqStmt,
    alterSeqStmt,
    createStmt,
    dropStmt,
    alterEnumStmt,
    alterOwnerStmt,
    indexStmt,
    alterTableStmt,
    selectStmt,
    viewStmt,
    renameStmt,
    updateStmt,

    // Standalone comments
    CommentStatement,
  ]),
  (v, ctx) => {
    if (!ctx.numStatements) {
      ctx.numStatements = 1;
    } else {
      ctx.numStatements++;
    }

    return v;
  }
);

export const stmts: Rule<Stmt[]> = transform(
  sequence([
    zeroToMany(stmt),
    endOfInput, // <-- This ensures we don't return before hitting the end of our SQL.
  ]),
  (v) => {
    const stmts = v[0];

    const output: Stmt[] = [];

    let previousStmtLocation = 0;
    for (let i = 0; i < stmts.length; i++) {
      const c = stmts[i];

      output.push({
        stmt: c.value,
        ...(c.eos?.firstSemicolonPos
          ? {
              stmt_len: c.eos.firstSemicolonPos - previousStmtLocation,
            }
          : {}),
        ...(previousStmtLocation
          ? { stmt_location: previousStmtLocation }
          : {}),
      });

      if (c.eos?.lastSemicolonPos != null) {
        previousStmtLocation = c.eos.lastSemicolonPos + 1;
      }
    }

    return output;
  }
);

function reduceComments(acc: Stmt[], stmt: Stmt): Stmt[] {
  const previousStmt = acc[acc.length - 1];

  // combine comment nodes and remove whitespace.
  if (
    previousStmt &&
    "Comment" in previousStmt.stmt &&
    "Comment" in stmt.stmt
  ) {
    return [
      ...acc.slice(0, -1),
      {
        stmt: {
          Comment: combineComments(
            previousStmt.stmt.Comment,
            stmt.stmt.Comment
          ),
        },
      },
    ];
    // Just lead whitespace
  } else if ("Comment" in stmt.stmt) {
    return [
      ...acc,
      {
        stmt: {
          Comment: stmt.stmt.Comment,
        },
      },
    ];
  } else {
    return acc.concat(stmt);
  }
}

/**
 * Used mostly of validation of parse method
 */
export function parseComments(inputSql: string) {
  const result = codeComments({
    str: inputSql,
    pos: 0,
    numStatements: 0,
  });

  if (result.type == ResultType.Success) {
    return result.value;
  }

  throw new Error("sql brokenz");
}

/**
 * Human-friendly version of the raw generated parser's parse() function, but
 * with much better error reporting, showing source line position where it
 * failed.
 */
export default function parse(
  context: Context,
  expectedAst: Stmt[] = []
): SuccessResult<Stmt[]> {
  const result = stmts(context);

  if (result.type == ResultType.Success) {
    result.value = result.value.reduce(reduceComments, []);
    return result;
  }

  const error = new ParseError(
    context,
    result,
    expectedAst[context.numStatements ?? 0]
  );

  throw error;
}
