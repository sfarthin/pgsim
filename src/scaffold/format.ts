import { resolve } from "path";
import { writeFileSync, readdirSync, existsSync } from "fs";
import prettier from "prettier";
import warning from "./warning";
import { stmtTypes, mapStmtTypes } from "./constants";
import { camelCase } from "lodash";

export default async function () {
  const parseStmtPath = resolve(__dirname, "../format/stmt.ts");
  const formatIndexPath = resolve(__dirname, "../format/index.ts");

  const files = readdirSync(resolve(__dirname, "../format/")).flatMap((f) =>
    f === "index.ts" ? [] : f.substring(0, f.length - 3)
  );

  const mapFiles = (mapFn: (n: string) => string) =>
    files.map(mapFn).join(`\n`);

  const importedFormatters = mapStmtTypes(
    (stmtName) =>
      `import ${camelCase(stmtName)} from "./${camelCase(stmtName)}";`
  );

  const exportedStmt = mapFiles((stmtName) => `export * from "./${stmtName}";`);

  const longConditional = stmtTypes
    .map(
      (stmtType) => /* ts */ `
    if ("${stmtType}" in s) {
      return ${camelCase(stmtType)}(s.${stmtType});
    }
  `
    )
    .join(` else `);

  writeFileSync(
    parseStmtPath,
    prettier.format(
      /* ts */ `

    ${warning}
    import { Stmt } from "~/types";
    import { Block } from "./util";
    ${importedFormatters}

    export default function stmt(stmt: Stmt): Block {
      const s = stmt.stmt;
    
      ${longConditional}
      
      throw new Error('Cannot format ' + Object.keys(s));
      
    }
    

    `,
      { parser: "babel" }
    )
  );

  writeFileSync(
    formatIndexPath,
    prettier.format(
      /* ts */ `

    ${warning}
    import format from "./format";
    export { toString } from "./print";
    export default format;
    ${exportedStmt}

    `,
      { parser: "babel" }
    )
  );

  for (const stmtType of stmtTypes) {
    const stmtFilePath = resolve(
      __dirname,
      `../format/${camelCase(stmtType)}.ts`
    );
    if (!existsSync(stmtFilePath)) {
      writeFileSync(
        stmtFilePath,
        prettier.format(
          /* ts */ `
        import { ${stmtType} } from "~/types";
        import { Block } from "./util";
        
        export default function (_c: ${stmtType}): Block {
          return [];
        }
      `,
          { parser: "babel" }
        )
      );
    }
  }
}
