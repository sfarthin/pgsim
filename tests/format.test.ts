import parse from "../src/parse";
import nParse from "../src/nativeParse";
import format from "../src/format";
import omitDeep from "./omitDeep";

function removeStyle(ast: object): object {
  return omitDeep(ast, ["location", "stmt_len", "stmt_location"]);
}

function expectFormatHasSameAst(sql: string): void {
  const ast1 = nParse(sql);
  const ast2 = parse(format(parse(sql)));

  expect(removeStyle(ast1)).toEqual(removeStyle(ast2));
}

describe("format", () => {
  it("create a table with all kinds of columns", () => {
    expectFormatHasSameAst(`
    CREATE TABLE a (
      -- integer
      aaa integer,
      aab int,
      aac int4,

      aa bigint,
      ab int8,
      ac bigserial,
      ad serial8,
      ae serial,
      af serial4,

      -- bit
      aba bit,
      abb bit(2),
      abc bit varying,
      abca bit varying(3),
      abd varbit,
      abda varbit(3),

      -- interval
      bba interval,
      bbb interval day to hour,

      -- boolean
      ba bool,
      bbaa boolean,

      -- misc
      caa box,
      cba bytea,
      cc cidr,
      cd circle,
      ce date,
      cf inet,
      cg line,
      ch lseg,
      ci macaddr,
      cj money,
      ck tsquery,
      cl tsvector,
      cm txid_snapshot,
      cn uuid,
      co xml,

      -- real
      daa real,
      dba float4,

      -- smallint
      dda smallint,
      ddb int2,

      -- float
      da double precision,
      db float8,

      -- decimal
      eaa decimal,
      eb numeric,
      eca decimal(5),
      ecb numeric(5),
      eccv decimal(5,2),
      ecd numeric(5,2),

      -- character
      bb character varying(255) NOT NULL,
      bc character varying(4),
      bd character varying,
      be varchar,
      bf varchar(255),
      bg character,
      bh character(4),
      bi char,
      bj char(4),
      bk text,

      -- timestamp
      caz timestamp with time zone NOT NULL,
      cbx timestamptz,
      cax time with time zone NOT NULL,
      cbxx timetz
    )
  `);
  });

  it("set vars", () => {
    expectFormatHasSameAst(`
      SET statement_timeout = 0;
      SET lock_timeout = 0;
      SET client_encoding = 'UTF8';
      SET standard_conforming_strings = on;
      SET check_function_bodies = false;
      SET client_min_messages = warning;
  `);
  });

  it("define enum", () => {
    expectFormatHasSameAst(`
      CREATE TYPE foo AS ENUM (
        'a',
        'b',
        'c',
        'd',
        'e'
      );`);
  });

  it("sequence", () => {
    expectFormatHasSameAst(`
    CREATE SEQUENCE foo
      START WITH 1
      INCREMENT BY 1
      NO MINVALUE
      NO MAXVALUE
      NO CYCLE
      CACHE 1
      OWNED BY NONE;

      CREATE SEQUENCE foo2
      MINVALUE 4
      MAXVALUE 6
      CYCLE
      OWNED BY foo.bar;
    `);
  });

  it("asd", () => {
    expectFormatHasSameAst(`ALTER SEQUENCE foo OWNED BY a.aaa;`);
  });

  it.only("can natively parse", () => {
    const sql = `
      CREATE SEQUENCE foo
      START WITH 1
      INCREMENT BY 1
      NO MINVALUE
      NO MAXVALUE
      NO CYCLE
      CACHE 1
      OWNED BY NONE;
    `;

    console.log(JSON.stringify(nParse(sql), null, 2));

    console.log(JSON.stringify(parse(sql), null, 2));
  });
});
