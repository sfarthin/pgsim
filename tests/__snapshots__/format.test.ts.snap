// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format asd 1`] = `
"ALTER SEQUENCE foo 
	OWNED BY a.aaa;"
`;

exports[`format define enum with comments 1`] = `
"-- Top 1

-- Top 2

-- 1
-- 2
-- 3
-- 4
-- 5
-- 6
-- 7
-- 8
CREATE TYPE foo AS ENUM (
	-- comment on a
	'a',
	-- comment on b
	'b',
	-- comment on c
	'c',
	-- comment on d
	'd',
	-- comment on e
	'e'
);

-- Bottom 1

-- Bottom 2
"
`;

exports[`format sequence 1`] = `
"CREATE SEQUENCE foo 
	START WITH 1
	INCREMENT BY 1
	NO MINVALUE
	NO MAXVALUE
	NO CYCLE
	CACHE 1
	OWNED BY NONE;
CREATE SEQUENCE foo2 
	MINVALUE 4
	MAXVALUE 6
	CYCLE
	OWNED BY foo.bar;"
`;
